### DISCUSSION TITLE
How to use LD_PRELOAD for all profiles?

### DISCUSSION THREAD

**Discussion by kmille** [netblue30/firejail#5512] [Q&A]
Hey,

I want to use `/usr/lib/libhardened_malloc.so` globally for all applications jailed by firejail. But I don't get it working with the `whitelist` or `private-lib` directives:

```bash
kmille@linbox:firejail cat globals.local    
env LD_PRELOAD=/usr/lib/libhardened_malloc.so
private-lib libhardened_malloc.so
```

This works for profiles that are using `private-lib` (like ping) or `private-lib gconv` (like whois). But it doesn't work for profiles that need full access to /lib like chromium or firefox.

Do you have an idea how to solve this problem? One can copy libhardened_malloc.so to /opt but I don't like this approach.

Reference: https://wiki.archlinux.org/title/Security#Hardened_malloc

**Comment by rusty-snake** [comment]
Commands in globlas.local will affect all profiles, so you can either

1. `ignore private-lib` in globales.local
2. or add `private-lib libhardened_malloc.so` to all relevant .locals (can be scripted).

Also note that firejail 0.9.72 will disable  private-lib by default.

**Comment by ghost** [comment]
> Also note that firejail 0.9.72 will disable private-lib by default.

Correct. _If_ you want to keep using it you'll need to override its setting in `/etc/firejail/firejail.config`:

```
[...]
# Enable or disable private-lib feature, default disabled
private-lib yes
[...]
```

I've been using hardened malloc with Firejail for a long time, and with good results. But I didn't implement it like you are trying. Mostly because IMO that can break applications rather easily if not handled carefully. I do have a `blacklist  /usr/lib/libhardened_malloc.so` in globals.local (so other profiles won't see it). Combining this with `noblacklist /usr/lib/libhardened_malloc.so` and `env LD_PRELOAD=/usr/lib/libhardened_malloc.so` in the profiles it is known to work and **make sense** this has proven quite solid/stable. Why do I stress `makes sense`? Well, because there are applications that simply won't work when having it in their sandbox, like Firefox (likely also Chromium):
- see #2689 for more context
- see [this](https://www.kicksecure.com/wiki/Hardened_Malloc#cite_ref-firefox-based_3-0) for Firefox
- see the Arch Wiki [here](https://wiki.archlinux.org/title/Firejail#Use_with_hardened_malloc) (although the example is confusing and contradicting the note)

Just my two cents.
